File: packages/pi-lean-portal/backends/python-base/pi_browser_bridge/bot_detection.py
"""
Bot detection for the Python bridge.

Mirrors the logic in ``core/shared/bot-detection.ts`` ``checkPage()``.

The matching *must* run in the Python process because it needs
``page.title()`` and ``page.evaluate()`` for body/HTML — shipping the
full HTML across JSON-RPC would be prohibitively expensive.  However,
the signal lists are pure data and are loaded from the shared
``browser-data.json`` file — the same file used by the TypeScript side
so the two implementations never drift.
"""

import re
from typing import Any

from .browser_data import BOT_SIGNALS

#: Block-level signals — checked against BOTH title and body text.
_BLOCK_SIGNALS: tuple[str, ...] = tuple(BOT_SIGNALS["blockSignals"])

#: Body-only string signals — high-specificity CDN patterns.
_BODY_ONLY_SIGNALS: tuple[str, ...] = tuple(BOT_SIGNALS["bodyOnlySignals"])

#: Body-only regex patterns — compiled from shared source strings.
_BODY_ONLY_PATTERNS: tuple[re.Pattern, ...] = tuple(
    re.compile(p, re.IGNORECASE) for p in BOT_SIGNALS["bodyOnlyPatterns"]
)

#: HTML-level CAPTCHA/widget signals.
_HTML_SIGNALS: tuple[str, ...] = tuple(BOT_SIGNALS["htmlSignals"])


def check_bot_detection(page: Any) -> bool:
    """Check for anti-automation / bot detection signals in the current page.

    Mirrors the logic in ``bot-detection.ts`` ``checkPage()``.

    Args:
        page: A Playwright sync Page object (or a mock with ``title()``
              and ``evaluate()`` methods).

    Returns:
        True if bot detection signals were found.
    """
    try:
        title: str = page.title().lower()
    except Exception:
        title = ""

    try:
        body_text: str = page.evaluate(
            "() => document.body?.innerText || ''"
        ) or ""
        body_text = body_text.lower()
    except Exception:
        body_text = ""

    try:
        html: str = page.evaluate(
            "() => document.documentElement?.innerHTML || ''"
        ) or ""
        html = html.lower()
    except Exception:
        html = ""

    # ── Block signals: checked against both title and body ───────
    for signal in _BLOCK_SIGNALS:
        if signal in title or signal in body_text:
            return True

    # ── Body-only string signals ────────────────────────────────
    for signal in _BODY_ONLY_SIGNALS:
        if signal in body_text:
            return True

    # ── Body-only regex patterns ────────────────────────────────
    for pattern in _BODY_ONLY_PATTERNS:
        if pattern.search(body_text):
            return True

    # ── HTML-level signals (Python-only enhancement) ────────────
    for signal in _HTML_SIGNALS:
        if signal in html:
            return True

    return False


__all__ = ["check_bot_detection"]

File: packages/pi-lean-portal/core/shared/bot-detection.ts
/**
 * Bot detection heuristics.
 *
 * Analyzes page content to determine if a site is blocking
 * automation tools (Cloudflare, CAPTCHA, etc.). When detected,
 * the router flags the navigation as bot-blocked so the agent
 * can decide how to proceed — try web-fetch, try a different URL,
 * or switch to a stealth browser backend if one is configured.
 *
 * Signal lists are loaded from the shared ``browser-data.json`` —
 * the same file used by the Python bridge so the two sides never drift.
 */

import { BOT_SIGNALS } from "./browser-data.js";

/**
 * Content patterns that indicate a bot block — loaded from shared data.
 */
const BLOCK_SIGNALS = BOT_SIGNALS.blockSignals;

/**
 * Body-only string patterns — loaded from shared data.
 */
const BODY_ONLY_SIGNALS = BOT_SIGNALS.bodyOnlySignals;

/**
 * Body-only regex patterns — compiled from shared data at module load.
 */
const BODY_ONLY_PATTERNS: RegExp[] = BOT_SIGNALS.bodyOnlyPatterns.map(
	(src) => new RegExp(src, "i"),
);

/**
 * HTML-level signals — loaded from shared data.
 */
const HTML_SIGNALS = BOT_SIGNALS.htmlSignals;

/**
 * Check if page text content suggests a bot block.
 */
function checkBodyText(bodyText: string): boolean {
	if (!bodyText) return false;

	const lower = bodyText.toLowerCase();
	for (const signal of BLOCK_SIGNALS) {
		if (lower.includes(signal)) {
			return true;
		}
	}

	return false;
}

/**
 * Check body text against body-only signal patterns that are specific
 * enough to not false-positive on normal content.  Checks both string
 * inclusion (for CDN domains, generic 403 messages) and regex patterns
 * (for Akamai reference codes, etc.).
 */
function checkBodyOnlyText(bodyText: string): boolean {
	if (!bodyText) return false;

	const lower = bodyText.toLowerCase();

	// Check string signals first
	for (const signal of BODY_ONLY_SIGNALS) {
		if (lower.includes(signal)) {
			return true;
		}
	}

	// Check regex patterns against raw text (case-insensitive via /i flag)
	for (const pattern of BODY_ONLY_PATTERNS) {
		const match = bodyText.match(pattern);
		if (match) {
			return true;
		}
	}

	return false;
}

/**
 * Check raw HTML content for CAPTCHA widget embed codes.
 *
 * Looks for specific widget identifiers in the HTML source (not visible text),
 * so false positive risk from generic mentions is low.  The signal list is
 * loaded from the shared ``browser-data.json`` so TypeScript and Python
 * bridge implementations share a single source of truth.
 */
function checkHtmlContent(html: string): boolean {
	if (!html) return false;

	const lower = html.toLowerCase();
	for (const signal of HTML_SIGNALS) {
		if (lower.includes(signal)) {
			return true;
		}
	}

	return false;
}

/**
 * Combined check: analyze page title, body text, and (optionally) raw HTML.
 *
 * - Title is checked against BLOCK_SIGNALS only (avoids false positives).
 * - Body is checked against BLOCK_SIGNALS + BODY_ONLY_SIGNALS + BODY_ONLY_PATTERNS
 *   (catches CDN-specific block pages like Akamai "Access Denied").
 * - HTML (if provided) is checked for CAPTCHA widget embed codes.
 */
export function checkPage(
	title: string,
	bodyText: string,
	html?: string,
): boolean {
	// Check title first (often contains "Attention Required!" etc.)
	if (checkBodyText(title)) return true;

	// Check body against challenge phrases
	if (checkBodyText(bodyText)) return true;

	// Check body against CDN-specific patterns (reference #, etc.)
	if (checkBodyOnlyText(bodyText)) return true;

	// Check HTML source for CAPTCHA widget embed codes
	if (html !== undefined && checkHtmlContent(html)) return true;

	return false;
}
